### PR TITLE
fix: support optional function parameters in cross-environment method calls

### LINKED ISSUE #1434: Services: 'undefined' in method arguments gets converted to null when calling a remote service
If you have the following service:
```ts
class EchoService {
    public echo(message = 'hello world') {
        return message;
    }
}
```

Calling it from a different environment as `EchoService.echo(undefined)` will return `null` instead of `'hello world'`.

This has tripped me a few times. Undefined should stay undefined.

### PR DESCRIPTION
**Pull request by vladyslav-baliuk**
Fixes #1434

### CODE REVIEW

**Comment by AviVahl** [comment]
Use stringify/parser replacer+reviver parameters instead?

File: packages/core/src/com/communication.ts [hunk 1]
```diff
@@ -240,13 +246,14 @@ export class Communication {
         envId: string,
         api: string,
         method: string,
-        args: unknown[],
+        unserializedArgs: unknown[],
```
Comment (alisey) [inline packages/core/src/com/communication.ts:249]: Just `args`

File: packages/core/src/com/communication.ts [hunk 2]
```diff
@@ -240,13 +246,14 @@ export class Communication {
...
         origin: string,
         serviceComConfig: Record<string, AnyServiceMethodOptions>,
         forwardingChain: string[]
     ): Promise<unknown> {
         return new Promise<void>((res, rej) => {
             const callbackId = !serviceComConfig[method]?.emitOnly ? this.idsCounter.next('c') : undefined;
+            const args = serializeApiCallArguments(unserializedArgs);
```
Comment (alisey) [inline packages/core/src/com/communication.ts:256]: Do it only if it's not a listen call.

File: packages/core/src/com/helpers.ts [hunk 3]
```diff
@@ -23,3 +23,11 @@ export class MultiCounter {
         return `${ns}${this.ids[ns]++}`;
     }
 }
+
+const undefinedArgPlaceholder = '__|UND_ARG_PLA|__';
```
Comment (alisey) [inline packages/core/src/com/helpers.ts:27]: undefinedPlaceholder = '__undefined_placeholder__';

File: packages/core/src/com/helpers.ts [hunk 4]
```diff
@@ -23,3 +23,11 @@ export class MultiCounter {
...
+
+export const serializeApiCallArguments = (args: unknown[]): unknown[] =>
```
Comment (alisey) [inline packages/core/src/com/helpers.ts:29]: Add a comment or a link to the issue explaining why this is necessary.

**Comment by alisey** [comment]
> fix: added serialization/deserialization for arguments of the API calls

This is a not a reasonable title for this pull request.

File: packages/core/test/node/com.spec.ts [hunk 5]
```diff
@@ -160,6 +160,25 @@ describe('Communication', () => {
         expect(handleMessageStub).to.have.not.been.called;
     });
 
+    it('serialization of arguments for remote API calls should work as expected ', async () => {
```
Comment (olehrakwix) [inline packages/core/test/node/com.spec.ts:163]: "should work as expected" does not really help to understand what is expected

File: test-fixtures/simple-feature/src/feature/y.feature.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { COM, Environment, Feature, Service } from '@wixc3/engine-core';
+
+export const serverEnv = new Environment('server', 'node', 'single');
+
+export default new Feature({
+    id: 'YTestFeature',
```
[the remark this answers is not in the document]
Comment (olehrakwix) [reply test-fixtures/simple-feature/src/feature/y.feature.ts:6]: yes, @vladyslav-baliuk , please rename feature and all files to something like `default-args-echo`

File: packages/core/src/com/communication.ts (hunk 1 above)
Comment (vladyslav-baliuk) [reply packages/core/src/com/communication.ts]: @alisey fixed

File: packages/core/src/com/communication.ts (hunk 2 above)
Comment (vladyslav-baliuk) [reply packages/core/src/com/communication.ts]: @alisey fixed

File: packages/core/src/com/helpers.ts (hunk 3 above)
Comment (vladyslav-baliuk) [reply packages/core/src/com/helpers.ts]: @alisey fixed

File: packages/core/src/com/helpers.ts (hunk 4 above)
Comment (vladyslav-baliuk) [reply packages/core/src/com/helpers.ts:33]: @alisey fixed

File: test-fixtures/simple-feature/src/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+      "outDir": "../dist"
+    },
+    "references": [
+      { "path": "../../../packages/core/src" },
+      {
+        "path": "../../../packages/core-node/src"
+      }
+    ]
+  }
+
```
Comment (olehrakwix) [inline test-fixtures/simple-feature/src/tsconfig.json:13]: please add newline at the end of the file

File: test-fixtures/simple-feature/src/feature/y.feature.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (vladyslav-baliuk) [reply test-fixtures/simple-feature/src/feature/y.feature.ts:6]: @alisey @olehrakwix it is just a simplified name for the feature that I'm using in test, actually the most of the features in `test-fixtures` have naming like that so I stayed consistent :)

File: packages/core/test/node/com.spec.ts (hunk 5 above)
Comment (vladyslav-baliuk) [reply packages/core/test/node/com.spec.ts]: @olehrakwix fixed

**Comment by vladyslav-baliuk** [comment]
> > fix: added serialization/deserialization for arguments of the API calls
> 
> This is a not a reasonable title for this pull request.

fixed

File: test-fixtures/simple-feature/src/feature/y.feature.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (olehrakwix) [reply test-fixtures/simple-feature/src/feature/y.feature.ts:6]: no, please still rename it so it has meaningful name

**Comment by barak007** [comment]
this work is going to affect all messages in COM. do we really need to support this behavior? 
can't we try a typed approach so whenever we see a argument or return of `undefined` for an async service we use `undefined | null`

**Comment by alisey** [comment]
> this work is going to affect all messages in COM. do we really need to support this behavior?

Yes, and what's the problem?

**Comment by barak007** [comment]
It's kind of a hot spot. We need to keep an eye that we don't do unnecessary work.

**Comment by alisey** [comment]
Vlad checked the impact on performance and didn't find any noticeable effect. (@vladyslav-baliuk , correct me if I'm wrong)

Additionally, we can do something like this:

```js
const args = [...];

function patchArguments(args) {
    for (let i = 0; i < args.length; i++) {
        if (args[i] === undefined) args[i] = undefinedPlaceholder;
    }
}
```

That way we wouldn't be creating extra work for the garbage collector. For most calls this would amount to merely looping through an array of 1-3 elements, which would amount to an overhead of ~0.00001 millisecond per remote API call.

But again, I'm not sure that would make any difference, modern generational garbage collector handle transient objects very efficiently.

File: test-fixtures/simple-feature/src/feature/y.feature.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (vladyslav-baliuk) [reply test-fixtures/simple-feature/src/feature/y.feature.ts]: @olehrakwix fixed

**Comment by vladyslav-baliuk** [comment]
> Vlad checked the impact on performance and didn't find any noticeable effect. (@vladyslav-baliuk , correct me if I'm wrong)
> 
> Additionally, we can do something like this:
> 
> ```js
> const args = [...];
> 
> function patchArguments(args) {
>     for (let i = 0; i < args.length; i++) {
>         if (args[i] === undefined) args[i] = undefinedPlaceholder;
>     }
> }
> ```
> 
> That way we wouldn't be creating extra work for the garbage collector. For most calls this would amount to merely looping through an array of 1-3 elements, which would add an overhead of ~0.00001 millisecond per remote API call. In other words, you would need to execute 100 000 remote API calls per second for this overhead to add up to a single millisecond.
> 
> But again, I'm not sure that would make any difference, modern generational garbage collector handle transient objects very efficiently.

Right, I've checked the amount of time needed for 

> Vlad checked the impact on performance and didn't find any noticeable effect. (@vladyslav-baliuk , correct me if I'm wrong)
> 
> Additionally, we can do something like this:
> 
> ```js
> const args = [...];
> 
> function patchArguments(args) {
>     for (let i = 0; i < args.length; i++) {
>         if (args[i] === undefined) args[i] = undefinedPlaceholder;
>     }
> }
> ```
> 
> That way we wouldn't be creating extra work for the garbage collector. For most calls this would amount to merely looping through an array of 1-3 elements, which would add an overhead of ~0.00001 millisecond per remote API call. In other words, you would need to execute 100 000 remote API calls per second for this overhead to add up to a single millisecond.
> 
> But again, I'm not sure that would make any difference, modern generational garbage collector handle transient objects very efficiently.

Correct, I've checked the amount of time needed to perform 100 remote API calls in a row, checked it on fresh machine several times per both of states - with and without serialization, no noticeable differences were found

Anyway I'll double-check it today with a bit different technique

File: packages/core/test/node/com.spec.ts [hunk 8]
```diff
@@ -160,6 +160,25 @@ describe('Communication', () => {
         expect(handleMessageStub).to.have.not.been.called;
     });
 
+    it('serialization of arguments for remote API calls should handle undefined values ', async () => {
```
Comment (alisey) [inline packages/core/test/node/com.spec.ts:163]: `supports default function parameters in remote API calls`

File: test-fixtures/default-args-echo/src/feature/feature.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { COM, Environment, Feature, Service } from '@wixc3/engine-core';
+
+export const serverEnv = new Environment('server', 'node', 'single');
+
+export default new Feature({
+    id: 'DefaultArgsEcho',
```
Comment (alisey) [inline test-fixtures/default-args-echo/src/feature/feature.ts:6]: camel case

File: test-fixtures/default-args-echo/src/feature/server.env.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import sampleFeature, { serverEnv } from './feature';
+
+sampleFeature.setup(serverEnv, () => {
+    return {
+        echoService: {
+            echo: (s = 'dude, it works!') => s,
```
Comment (alisey) [inline test-fixtures/default-args-echo/src/feature/server.env.ts:6]: Why do you need two echo services, `echo(s = 'No!')` and `echo(s = 'dude, it works!')`

File: packages/core/test/node/com.spec.ts (hunk 8 above)
Comment (vladyslav-baliuk) [reply packages/core/test/node/com.spec.ts]: @alisey fixed

File: test-fixtures/default-args-echo/src/feature/feature.ts (hunk 9 above)
Comment (vladyslav-baliuk) [reply test-fixtures/default-args-echo/src/feature/feature.ts]: @alisey fixed

File: test-fixtures/default-args-echo/src/feature/server.env.ts (hunk 10 above)
Comment (vladyslav-baliuk) [reply test-fixtures/default-args-echo/src/feature/server.env.ts]: @alisey the first one is for testing serialization/deserialization itself but actually it should be without default value, just optional, fixed

The second one is a part of the feature created to test serialization in real remote API calls

Comment (alisey) [reply test-fixtures/default-args-echo/src/feature/server.env.ts]: I still don't understand why we need the first one if the second one is testing a real remote API call.

Comment (vladyslav-baliuk) [reply test-fixtures/default-args-echo/src/feature/server.env.ts]: @alisey the idea was to cover that functionality in two test-suits  - as part of `communication` tests to verify only serialization/deserialization process and as a part of `runtime-node` tests to verify how it works in real communication between envs

But you are right - the second test-case is pretty enough, fixed

File: test-fixtures/default-args-echo/src/feature/echo.server.env.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import sampleFeature, { serverEnv } from './echo.feature';
```
Comment (alisey) [inline test-fixtures/default-args-echo/src/feature/echo.server.env.ts:1]: Rename the import.

**Pull request merged**